### PR TITLE
Update the minimum external LLVM to 16.

### PR DESCRIPTION
**Pull request by Dirbaio**
With this change, we'll have stable support for LLVM 16 and 17.
For reference, the previous increase to LLVM 15 was #114148

[Relevant zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/riscv.20forced-atomics)

### CODE REVIEW

File: compiler/rustc_target/src/spec/mod.rs [hunk 1]
```diff
@@ -1155,10 +1155,6 @@ pub enum StackProbeType {
 }
 
 impl StackProbeType {
-    // LLVM X86 targets (ix86 and x86_64) can use inline-asm stack probes starting with LLVM 16.
-    // Notable past issues were rust#83139 (fixed in 14) and rust#84667 (fixed in 16).
-    const X86: Self = Self::InlineOrCall { min_llvm_version_for_inline: (16, 0, 0) };
-
```
Comment (cuviper) [inline compiler/rustc_target/src/spec/mod.rs:1161]: In some followup, we might want to remove the `__rust_probestack` stuff altogether, but that doesn't need to happen now.

File: tests/codegen/stack-probes-inline.rs [hunk 2]
```diff
@@ -13,10 +13,8 @@
 //[s390x] needs-llvm-components: systemz
 //[i686] compile-flags: --target i686-unknown-linux-gnu
 //[i686] needs-llvm-components: x86
-//[i686] min-llvm-version: 16
 //[x86_64] compile-flags: --target x86_64-unknown-linux-gnu
 //[x86_64] needs-llvm-components: x86
-//[x86_64] min-llvm-version: 16
```
Comment (cuviper) [inline tests/codegen/stack-probes-inline.rs:19]: Related to this file, `tests/codegen/stack-probes-call.rs` is now out of scope.

**Review by cuviper — APPROVED** [review]
I'm good with this update both as a wg-llvm member and downstream for Fedora compatibility. I'll still give @davidtwco a chance to review though, in case he sees any reason to hold on to LLVM 15 longer.

There's one more version test that can be removed here:

```
tests/ui/numbers-arithmetic/issue-105626.rs:3:// min-system-llvm-version: 16
```

File: tests/codegen/stack-probes-inline.rs (hunk 2 above)
Comment (Dirbaio) [reply tests/codegen/stack-probes-inline.rs:19]: Removed!

**Comment by Dirbaio** [comment]
> There's one more version test that can be removed here:

removed.

**Comment by klensy** [comment]
Looking at diff between x86_64-gnu-llvm-15 and x86_64-gnu-llvm-16 dockerfiles, there was things added in https://github.com/rust-lang/rust/pull/117313, they no longer needed in 16?

Then `USE_NEW_MANGLING` env can be cleaned up.

P.S. Is codegen_gcc tests even run? Can't find it in logs.

**Comment by cuviper** [comment]
Sorry! I missed that difference -- would you mind opening a PR to restore that bit?

**Comment by cuviper** [comment]
I opened #118463 to restore (and hopefully fix) those tests.

**Pull request merged**